base/server/src/main/java/org/dogtagpki/server/rest/v2/UserServletBase.java

```diff
@@ -0,0 +1,1146 @@
+
+                    X509Certificate[] p7certs = pkcs7.getCertificates();
+
+                    if (p7certs.length == 0) {
+                        logger.error("UserServletBase: PKCS #7 data contains no certificates");
+                        throw new BadRequestException("PKCS #7 data contains no certificates");
+                    }
+
+                    // fix for 370099 - cert ordering can not be assumed
+                    // find out the ordering ...
+
+                    // self-signed and alone? take it. otherwise test
+                    // the ordering
+                    if (p7certs[0].getSubjectDN().toString().equals(
+                            p7certs[0].getIssuerDN().toString()) &&
+                            (p7certs.length == 1)) {
+                        cert = p7certs[0];
+                        logger.debug("UserServletBase: {}", CMS.getLogMessage("ADMIN_SRVLT_SINGLE_CERT_IMPORT"));
+
+                    } else if (p7certs[0].getIssuerDN().toString().equals(p7certs[1].getSubjectDN().toString())) {
+                        cert = p7certs[0];
+                        logger.debug("UserServletBase: {}", CMS.getLogMessage("ADMIN_SRVLT_CERT_CHAIN_ACEND_ORD"));
+
+                    } else if (p7certs[1].getIssuerDN().toString().equals(p7certs[0].getSubjectDN().toString())) {
+                        assending = false;
+                        logger.debug("UserServletBase: {}", CMS.getLogMessage("ADMIN_SRVLT_CERT_CHAIN_DESC_ORD"));
+                        cert = p7certs[p7certs.length - 1];
+
+                    } else {
+                        // not a chain, or in random order
+                        logger.error("UserServletBase: {}", CMS.getLogMessage("ADMIN_SRVLT_CERT_BAD_CHAIN"));
+                        throw new BadRequestException(CMS.getUserMessage(loc, "CMS_USRGRP_SRVLT_CERT_ERROR"));
+                    }
+
+                    logger.debug("UserServletBase: {}",
+                            CMS.getLogMessage("ADMIN_SRVLT_CHAIN_STORED_DB", String.valueOf(p7certs.length)));
+
+                    int j = 0;
+                    int jBegin = 0;
+                    int jEnd = 0;
+
+                    if (assending) {
+                        jBegin = 1;
+                        jEnd = p7certs.length;
+                    } else {
+                        jBegin = 0;
+                        jEnd = p7certs.length - 1;
+                    }
+
+                    // store the chain into cert db, except for the user cert
+                    for (j = jBegin; j < jEnd; j++) {
+                        logger.debug("UserServletBase: {}",
+                                CMS.getLogMessage("ADMIN_SRVLT_CERT_IN_CHAIN", String.valueOf(j),
+                                        String.valueOf(p7certs[j].getSubjectDN())));
+                        org.mozilla.jss.crypto.X509Certificate leafCert =
+                                manager.importCACertPackage(p7certs[j].getEncoded());
```

Comment:
This is an existing code, but I don't think we should be importing the cert chain into the server's NSS database. We probably should remove it in a separate PR.

Reply:
Yes, this could be modified in a separate PR when we have fully tested the new APIs.